src/test/CharterManager-unit.t.sol

```diff
@@ -0,0 +1,564 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+import "./TestBase.sol";
+import {ManagedGemJoin} from "lib/dss-gem-joins/src/join-managed.sol";
+import "src/CharterManager.sol";
+
+contract MockVat {
+    struct Urn {
+        uint256 ink;   // Locked Collateral  [wad]
+        uint256 art;   // Normalised Debt    [wad]
+    }
+    struct Ilk {
+        uint256 Art;   // Total Normalised Debt     [wad]
+        uint256 rate;  // Accumulated Rates         [ray]
+        uint256 spot;  // Price with Safety Margin  [ray]
+        uint256 line;  // Debt Ceiling              [rad]
+        uint256 dust;  // Urn Debt Floor            [rad]
+    }
+    mapping (bytes32 => Ilk) public ilks;
+    mapping (bytes32 => mapping (address => uint256)) public gem;
+    mapping (bytes32 => mapping (address => Urn)) public urns;
+    mapping (address => uint256) public dai;
+    uint256 public live = 1;
+
+    function mockIlk(bytes32 ilk, uint256 _rate) external {
+        ilks[ilk].rate = _rate;
+    }
+    function add(uint256 x, int256 y) internal pure returns (uint256 z) {
+        z = x + uint256(y);
+        require(y >= 0 || z <= x, "vat/add-fail");
+        require(y <= 0 || z >= x, "vat/add-fail");
+    }
+    function sub(uint256 x, int256 y) internal pure returns (uint256 z) {
+        z = x - uint256(y);
+        require(y <= 0 || z <= x, "vat/sub-fail");
+        require(y >= 0 || z >= x, "vat/sub-fail");
+    }
+    function add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "vat/add-fail");
+    }
+    function sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "vat/sub-fail");
+    }
+    function slip(bytes32 ilk, address usr, int256 wad) external {
+        gem[ilk][usr] = add(gem[ilk][usr], wad);
+    }
+    function frob(bytes32 ilk, address u, address v, address w, int256 dink, int256 dart) external {
```

Comment:
I think MockVat should enforce delegation via `vat.hope()`. Currently the tests repaying loans should fail without the user doing `vat.hope(manager)`. Maybe this is more for integration tests, but it's small enough I think we can include it here. I'm okay if you leave it out though for the unit tests.

Reply:
done